File: H1_SDK1.3.6/camera_sdk_cpp/examples/02_minimal_depth.cpp
/**
 * @file 02_minimal_depth.cpp
 * @brief 【示例 02】深度图原始数据读取与可视化
 * * [功能介绍]
 * 获取 16位 (uint16_t) 原始深度数据并展示像素读取与伪彩色转换。
 * * [使用说明]
 * 1. 修改 GRPC_TARGET 为服务端真实的 IP 地址和端口（例如 "192.168.1.100:50051"）。
 * 2. 修改 CAMERA_NAME 为服务端实际挂载的相机 ID（如 "rs/cam_left_wrist"）。
 *  *[注意]
 * 如果连接失败，请优先检查网络是否互通（ping）以及防火墙是否放行了 50051 端口。
 */

#include <iostream>
#include <opencv2/opencv.hpp>
#include "camera_client.h"

int main() {
  // ---------------------------------------------------------
  // 【用户配置区】请根据实际环境修改以下参数
  // ---------------------------------------------------------
  const std::string GRPC_TARGET = "localhost:50051";    // 示例：修改为真实 IP
  const std::string CAMERA_NAME = "rs/cam_left_wrist";  // 深度相机 ID
  // ---------------------------------------------------------

  CameraClient client(GRPC_TARGET);
  client.start();

  while (true) {
    auto depth_data = client.getLatestDepth(CAMERA_NAME);
    if (depth_data) {
      cv::Mat depth_raw = depth_data->first;

      // 可视化处理
      cv::Mat depth_viz;
      depth_raw.convertTo(depth_viz, CV_8U, 0.03);
      cv::applyColorMap(depth_viz, depth_viz, cv::COLORMAP_JET);
      cv::imshow("Color Viewer", depth_viz);
    }

    if (cv::waitKey(1) == 'q') break;
  }

  client.stop();
  return 0;
}

File: H1_SDK_1.3.7/camera_sdk_python/example/02_minimal_depth.py
"""
@file 02_minimal_depth.py
@brief 【示例 02】深度图原始数据读取与可视化
* [功能介绍]
1. 展示如何获取 16位 (uint16) 原始深度像素数据。
2. 演示如何读取特定像素点（如中心点）的物理深度值（单位：mm）。
3. 展示将原始深度数据映射为伪彩色图（Color Map）以进行预览。
* [使用说明]
1. 修改 GRPC_TARGET 为服务端 IP，确保相机支持深度流。
2. 原始深度图直接显示通常全黑，必须经过缩放转换。
* [关键技术]
- 伪彩色应用：使用 cv2.COLORMAP_JET 增强深度层次感。
"""

import os
import sys
import cv2
import numpy as np

# 路径兼容：将 src 目录加入搜索路径以导入 camera_client
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
from camera_client import CameraClient

def main():
    GRPC_TARGET = "localhost:50051"
    CAMERA_NAME = "rs/cam_left_wrist"

    client = CameraClient(grpc_target=GRPC_TARGET)
    client.start()

    print(f"[*] 深度流取流中: {CAMERA_NAME}，按 'q' 键退出...")

    try:
        while True:
            # 获取数据 (numpy 数组格式)
            data = client.get_latest_depth(CAMERA_NAME)
            
            if data:
                depth_raw, timestamp = data
                
                # 可视化处理
                # 将 16-bit 线性缩放到 8-bit (0.03 为经验值，可根据实际量程调整)
                depth_viz = (depth_raw * 0.03).clip(0, 255).astype(np.uint8)
                # 应用色彩映射
                depth_color = cv2.applyColorMap(depth_viz, cv2.COLORMAP_JET)

                cv2.imshow("Depth Viewer", depth_color)

            if cv2.waitKey(1) & 0xFF == ord('q'):
                break
    finally:
        client.stop()
        cv2.destroyAllWindows()

if __name__ == "__main__":
    main()